### PR TITLE
REF: ignore_failures in BlockManager.reduce

### PR DESCRIPTION
**Pull request by jbrockmendel**
Moving towards collecting all of the ignore_failures code in one place.

The case where we have object dtypes is kept separate in this PR, will be handled in the next pass.

### CODE REVIEW

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -330,18 +330,31 @@ def _verify_integrity(self) -> None:
                 f"tot_items: {tot_items}"
             )
 
-    def reduce(self: T, func) -> T:
+    def reduce(self: T, func, ignore_failures: bool = False) -> Tuple[T, np.ndarray]:
```
Comment (jreback) [inline pandas/core/internals/managers.py:333]: can you add a doc-string, ideally type func

Comment (jbrockmendel) [reply pandas/core/internals/managers.py]: updated+green

File: pandas/core/frame.py [hunk 2]
```diff
@@ -8630,7 +8632,9 @@ def _get_data(axis_matters):
                 raise NotImplementedError(msg)
             return data
 
-        if numeric_only is not None and axis in [0, 1]:
+        if (numeric_only is not None and axis in [0, 1]) or (
```
Comment (jreback) [inline pandas/core/frame.py:8635]: i think this is worth a comment about what is being selected here as this is a complicated expression.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -8648,12 +8653,15 @@ def blk_func(values):
 
             # After possibly _get_data and transposing, we are now in the
             #  simple case where we can use BlockManager.reduce
-            res = df._mgr.reduce(blk_func)
-            out = df._constructor(res,).iloc[0].rename(None)
+            res, indexer = df._mgr.reduce(blk_func, ignore_failures=ignore_failures)
+            out = df._constructor(res).iloc[0]
```
Comment (jreback) [inline pandas/core/frame.py:8657]: we don't need to rename?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/frame.py:8635]: will do

File: pandas/core/frame.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/frame.py:8657]: changed the Index in BlockManager.reduce from `Index([0])` to `Index([None])` to make the rename unnecessary

File: pandas/core/frame.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/frame.py:8635]: updated+green

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_analytics.py [hunk 4]
```diff
@@ -1108,10 +1108,10 @@ def test_any_all_bool_only(self):
                 True,
                 marks=[td.skip_if_np_lt("1.15")],
             ),
-            (np.all, {"A": pd.Series([0, 1], dtype="category")}, False),
-            (np.any, {"A": pd.Series([0, 1], dtype="category")}, True),
+            (np.all, {"A": pd.Series([0, 1], dtype="category")}, True),
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/tests/frame/test_analytics.py:1111]: Not a bugfix per se, this is the behavior that changes if we declare that `ser.to_frame().all()` should be consistent with `ser.all()`, xref #36076

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/tests/frame/test_analytics.py:1071]: ok i think we need to deprecate this right (that was consensus?) 

also i suppose ok to just change. this is not a very large case. cc @jorisvandenbossche

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/tests/frame/test_analytics.py:1071]: updated with whatsnew

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/tests/frame/test_analytics.py:1071]: You have another PR actually trying to deprecate this right?

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/tests/frame/test_analytics.py:1071]: Also in https://github.com/pandas-dev/pandas/issues/36076, you commented a few days ago "the consensus seems to be that we should deprecate the current behavior in favor of matching the Series behavior". But so this PR is not doing that? Or is this PR not handling that case?

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/tests/frame/test_analytics.py:1071]: Closed that one.  This is a giant PITA and we should just rip the bandaid off.

**Review by jorisvandenbossche — COMMENTED** [review]
Can you try to summarize the issue a bit more? 

So with this PR, a categorical column will be skipped for `any`/`all` (does this also impact other reductions? Or other dtypes? Are there tests for this?). 
This is because with this PR it now takes the route of BlockManager.reduce->Block.reduce where Categorical._reduce is called, which raises an error, and BlockManager.reduce catches this error and skips the column. Is that correct? 
In master, on the other hand, for `any`/ `all` (filter reductions) we convert the full dataframe to an array (at least with the default `numeric_only=None`, and then on the numpy array the any/all operation works.

But so, that also means that the behaviour with this PR depends on the presence of (another) object dtype column or not? 

If we want to deprecate this, would a relatively clean option be: pass through the name of the op to `Block.reduce`, and let CategoricalBlock.reduce have a special case checking for the any/all op (in which case we can raise a warning and perform the op on the ndarray),  and otherwise use the normal Block.reduce implementation ?

File: doc/source/whatsnew/v1.2.0.rst [hunk 5]
```diff
@@ -244,7 +244,7 @@ Timezones
 
 Numeric
 ^^^^^^^
--
+- Bug in :class:`DataFrame` reductions incorrectly ignoring ``ExtensionArray`` behaviors (:issue:`35881`)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply doc/source/whatsnew/v1.2.0.rst:247]: Mostly for single-column EA-dtypes where the reduction on the array would raise.  Tough to put into a succinct note bc of the dropping-failures behavior.  suggestions welcome

**Comment by jbrockmendel** [comment]
> But so, that also means that the behaviour with this PR depends on the presence of (another) object dtype column or not?

I think it means the behavior in _master_ depends on the presence of (another) object dtpye column.  Which is part of the cluster-frack that is the reason we should rip off the bandaid.

**Comment by jbrockmendel** [comment]
Reverted behavior-changing component so that bug can be fixed separately

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. i know you are trying to cleanup this whole area, so good to go. @jorisvandenbossche any comments (I am sure this is going to be simplified over time)

**Comment by jbrockmendel** [comment]
yah the real simplifications dont come until we do the actual bugfixes, but this is a step in the right direction

**Comment by jreback** [comment]
great. glad finally to get this in.

File: pandas/core/frame.py [hunk 6]
```diff
@@ -8595,6 +8595,7 @@ def _reduce(
             cols = self.columns[~dtype_is_dt]
             self = self[cols]
 
+        any_object = self.dtypes.apply(is_object_dtype).any()
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py:8598]: I think this is partly the culprit of the slowdown. See also the top post of https://github.com/pandas-dev/pandas/pull/33252, which shows that `self.dtypes.apply(..)` is slower than the method that is used a few lines above for `dtype_is_dt`

**Review by jorisvandenbossche — COMMENTED** [review]
Doing some profiling on the example from https://github.com/pandas-dev/pandas/issues/37081, and reporting a few findings here.

File: pandas/core/frame.py [hunk 7]
```diff
@@ -8621,22 +8622,36 @@ def _get_data() -> DataFrame:
...
                 axis = 0
 
+            ignore_failures = numeric_only is None
+
             # After possibly _get_data and transposing, we are now in the
             #  simple case where we can use BlockManager.reduce
-            res = df._mgr.reduce(blk_func)
-            out = df._constructor(res).iloc[0].rename(None)
+            res, indexer = df._mgr.reduce(blk_func, ignore_failures=ignore_failures)
+            out = df._constructor(res).iloc[0]
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/frame.py:8648]: The getitem being `iloc[0]`?

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/core/frame.py:8648]: Indeed

**Pull request merged**